[ECOM-7345] User Credential API endpoint should be site aware

marjev opened the pull request:
@mjfrey @vkaracic @iivic This is the PR for making the User Credential API endpoint (WL) site aware.

The only reason this is marked as WIP is the fact that we still don't know to what extent should the endpoint be site-aware. I made it absolutely site-aware, which means that it doesn't matter which user sends the request to the endpoint, but from which Site the request is being sent from.
If we decide to remove the Site awareness for a specific use case, this PR can easily be modified.

JIRA ticket: https://openedx.atlassian.net/browse/ECOM-7345

vkaracic commented on credentials/apps/credentials/tests/factories.py:
```diff
@@ -51,6 +52,16 @@ class Meta(object):
+        """ Create a batch of UserCredential objects from the same site. """
+        user_credentials = []
+        for _ in range(size):
+            site = kwargs.get('site', Site.objects.get_current())
```
You don't need to get the site in every iteration.

vkaracic commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -225,39 +234,46 @@ def assert_list_username_filter_request_succeeds(self, username, expected):
         """ Verify the endpoint returns data for all UserCredentials awarded to the user matching the username. """
 
-        UserCredentialFactory.create_batch(3)
+        UserCredentialFactory.create_batch(3, credential=self.current_site_program_certificate)
```
Doesn't ``create_site_aware_batch`` do this automatically?

marjev replied:
There's a subtle difference.
`create_batch` will create n UserCredentials pointing to the same credential. (1 to many)
`create_site_aware_batch` will create both n `UserCredentials` and `ProgramCertificates`. (1 to 1).

vkaracic replied:
I don't think so: https://github.com/FactoryBoy/factory_boy/blob/v2.8.1/factory/base.py#L626

clintonb replied:
All you need here is `credential__site`. The specific credential doesn't matter.

vkaracic commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -273,3 +289,16 @@ def test_update(self, method):
+        UserCredentialFactory(credential=ProgramCertificateFactory(site=SiteFactory()))
+
+        self.authenticate_user(self.user)
+        self.add_user_permission(self.user, 'view_usercredential')
```
Doesn't concern your changes, but shouldn't the user be able to view user credentials by default?

marjev replied:
Only the ones (s)he owns, AFAIK.

vkaracic commented on credentials/apps/api/v2/views.py:
```diff
@@ -15,9 +16,21 @@ class CredentialViewSet(viewsets.ModelViewSet):
+        site = self.request.site
+        return UserCredential.objects.filter(
+            Q(program_credentials__site=site) |
+            Q(course_credentials__site=site)
```
Nit: this can fit in one line.

marjev commented:
Comments addressed.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -28,6 +29,7 @@ class CredentialViewSetTests(APITestCase):
 
     def setUp(self):
         super(CredentialViewSetTests, self).setUp()
+        self.current_site_program_certificate = ProgramCertificateFactory(site=Site.objects.get_current())
```
The factory already associates a `Site` with `ProgramCertificate` instances: https://github.com/edx/credentials/blob/352cb74284e38281977e68ebab9537cdb432703f/credentials/apps/credentials/tests/factories.py#L23. If instances need to be associated with a specific `Site`, create `self.site = SiteFactory()` and use `self.site` where needed.

marjev replied:
I want to use the current `Site` that the `Client` request will be associated with. I tried to use `SiteFactory` but I didn't find a way to specify which `Site` should be used by the middleware.
Do you know of a way to modify the `Client` to accomplish that?
I don't want to mock the request using the `RequestFactory` and call the view directly.

As a short term solution, I'll do the following (similar to what's done in the ecommerce within the `SiteMixin`):
```
Site.objects.all().delete()
self.site = SiteFactory(id=settings.SITE_ID)
```

clintonb replied:
https://github.com/edx/ecommerce/blob/master/ecommerce/tests/mixins.py#L250-L266

clintonb replied:
I would avoid relying on `settings.SITE_ID` if possible. That just seems brittle.

marjev replied:
The problem I'm having is that I don't know how to specify a site created by the `SiteFactory` as the site for middleware to use.
Almost every result of my Google search suggests using the `RequestFactory` for this, but that would mean I'd have to call the view directly as a method and not rely on Django's middleware and routing.

Regarding the URL you provided, I saw that, but do we need `SiteConfiguration` since it's not used in the `CredentialViewSetTests`?

clintonb replied:
The middleware uses the domain name. You only need the `Site` model. Note that the arguments for the `Site` model are prefixes with `site__` since it is a foreign key. See the following lines for the domain:

* https://github.com/edx/ecommerce/blob/master/ecommerce/tests/mixins.py#L252
* https://github.com/edx/ecommerce/blob/master/ecommerce/tests/mixins.py#L264

marjev replied:
I know what the middleware uses once it has the site, but I don't know how to specify a site to be used by the middleware. Removing `settings.SITE_ID` throws `Site DoesNotExist` my way, because of this https://github.com/django/django/blob/master/django/contrib/sites/models.py#L58. Someone specified `SITE_ID` in the `base.py` and it's being used to get the current site for middleware to use.

Also, I was asking why do we need to indirectly create a Site through `SiteConfigurationFactory`, since we're not using `SiteConfiguration` in this `TestCase`. Can't we simply create a site using the `SiteFactory`?

clintonb replied:
You don't *need* to use SiteConfiguration. I was simply pointing to an example. That being said the system is setup such that every Site has a SiteConfiguration. You may run into issues if the view expects a SiteConfiguration and one does not exist.

marjev replied:
I think it's addressed.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -205,8 +214,8 @@ def test_list(self):
         """ Verify the endpoint returns data for all UserCredentials that match the specified status. """
-        awarded = UserCredentialFactory.create_batch(3, status=UserCredential.AWARDED)
-        revoked = UserCredentialFactory.create_batch(3, status=UserCredential.REVOKED)
+        awarded = UserCredentialFactory.create_site_aware_batch(3, status=UserCredential.AWARDED)
```
The new method is not necessary. Simply pass the site on the `credential__site` kwarg.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -225,39 +234,46 @@ def assert_list_username_filter_request_succeeds(self, username, expected):
         self.authenticate_user(self.user)
         self.add_user_permission(self.user, 'view_usercredential')
 
-        response = self.client.get(self.list_path + '?program_uuid={}'.format(program_certificate.program_uuid))
```
Revert, and modify `program_certificate`.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -225,39 +234,46 @@ def assert_list_username_filter_request_succeeds(self, username, expected):
         UserCredentialFactory.create_batch(3)
-        program_certificate = ProgramCertificateFactory()
-        expected = UserCredentialFactory.create_batch(3, credential=program_certificate)
+        expected = UserCredentialFactory.create_batch(3, credential=self.current_site_program_certificate)
```
Revert, and modify `program_certificate`.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -273,3 +289,16 @@ def test_update(self, method):
         self.assertEqual(response.status_code, 200)
         self.assertEqual(response.data, self.serialize_user_credential(credential))
+
+    def test_get_queryset(self):
```
`test_site_filtering`

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -273,3 +289,16 @@ def test_update(self, method):
         self.assertEqual(response.data, self.serialize_user_credential(credential))
+
+    def test_get_queryset(self):
+        """ Verify the endpoint is site-aware. """
```
Verify the endpoint only returns credentials linked to a single site.

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -273,3 +289,16 @@ def test_update(self, method):
+        self.add_user_permission(self.user, 'view_usercredential')
+
+        response = self.client.get(self.list_path)
+        self.assertEqual(UserCredential.objects.all().count(), 2)
```
No need for `.all()`.

clintonb commented on credentials/apps/api/v2/views.py:
```diff
@@ -15,9 +16,18 @@ class CredentialViewSet(viewsets.ModelViewSet):
     serializer_class = UserCredentialSerializer
 
+    def get_queryset(self):
+        """ Get queryset containing only UserCredentials that belong to the current site.
```
No need for the docstring. The parent method already has one.

clintonb replied:
Add an inline comment instead: We have to filter on the explicit credential models because we cannot set a GenericRelation field on the Site model.

clintonb commented on credentials/apps/credentials/tests/factories.py:
```diff
@@ -51,6 +52,16 @@ class Meta(object):
     uuid = factory.LazyFunction(uuid.uuid4)
 
+    @classmethod
+    def create_site_aware_batch(cls, size, **kwargs):
```
This is not needed.

clintonb commented on credentials/apps/api/v2/views.py:
```diff
@@ -15,9 +16,18 @@ class CredentialViewSet(viewsets.ModelViewSet):
+            - CourseCredential
+        """
+        site = self.request.site
+        return UserCredential.objects.filter(Q(program_credentials__site=site) | Q(course_credentials__site=site))
```
You need to make use of the super-class:

```python
def get_queryset(self):
    site = self.request.site
    qs = super(CredentialViewSet, self).get_queryset()

    # We have to filter on the explicit credential models because we cannot set a GenericRelation
    # field on the Site model.
    qs = qs.filter(Q(program_credentials__site=site) | Q(course_credentials__site=site))

    return qs
```

marjev commented:
@clintonb, I've addressed your comments. Can you take a look at the 
[`setUp` implementation](https://github.com/edx/credentials/pull/179/files#diff-4ddad45f4c95e4fa7ed3a89da2f04116R33) because I would make sure it's acceptable to do it like that? Thanks!

clintonb commented on credentials/apps/api/v2/tests/test_views.py:
```diff
@@ -28,11 +30,22 @@ class CredentialViewSetTests(APITestCase):
+        domain = 'testserver.fake'
+        self.client = self.client_class(SERVER_NAME=domain)
+
+        Site.objects.all().delete()
```
Is this still necessary? If so, please add a comment explaining why.

clintonb commented:
Remember to squash your commits.

marjev commented:
@clintonb While I was exploring the credentials repo, I ran into `SiteMixin` in the most expecting place, core app. So, I used that and removed everything from the `setUp` method.
